### PR TITLE
libffi: Add 3.4.6, remove 3.4.2 and 3.4.3

### PR DESCRIPTION
**Pull request by Ahajha**
Specify library name and version:  **libffi/3.4.6**

The patch `0007-3.4.3-forward-declare-open_temp_exec_file.patch` (and friends) is no longer needed for the new version, as it has been upstreamed.

The MSVC build changed slightly: https://github.com/libffi/libffi/commit/2b2f61f6e6a79d46fa277965745a8cd8bb5f04bc, essentially there's a new `FFI_STATIC_BUILD` that replaces `FFI_BUILDING`.

I'm not certain that the `FFI_BUILDING`/`FFI_STATIC_BUILD` macros need to be in the package info for non-MSVC builds. I've kept them for now, but maybe we could revisit at some point.
 
Older versions of libffi don't support Apple ARM CPUs, it looks like there was an attempt to backport the fix, but the patch itself wasn't actually merged: https://github.com/conan-io/conan-center-index/pull/5085. There are plenty of newer versions so I don't think it's worth fixing at this point. Unsure why this was still an issue.

---

- [X] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [X] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [X] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by Ahajha** [comment]
@uilianries @RubenRBS Do you think it's about time to remove libffi 3.4.2 and 3.4.3? 3.4.4 is used everywhere (except llvm-core which is still on 3.3). 3.4.4 is already about two years old, I don't think we need anything older.

**Comment by Ahajha** [comment]
...Very confused as to why 3.3 broke. Will investigate.

**Review by RubenRBS — DISMISSED** [review]
Thanks a lot for the detailed PR body, it's super appreciated :)

**Comment by Ahajha** [comment]
@RubenRBS What do you think about removing the older versions? Only 3.3 is used, and it's by llvm-core, which is getting updated in another PR. 3.4.4 itself has been out for nearly 2 years, I don't think we need anything older.

**Comment by uilianries** [comment]
@Ahajha you could remove `3.4.3` and `3.4.2`. Others I would prefer in another PR, after updating other recipes first. Thank you for checking it.

**Comment by Ahajha** [comment]
@uilianries I've removed those two versions. I realize my comment wasn't clear, 3.4.4 I think should stay for now (and would have to anyways), 3.3 was the other version that I think should be removed, but I'll wait until the llvm-core PR is merged to do that.

**Comment by uilianries** [comment]
@Ahajha You need to remove both version from conandata.yml too.

**Comment by Ahajha** [comment]
@uilianries Thanks, was a bit tired last night :)

**Comment by uilianries** [comment]
> @uilianries Thanks, was a bit tired last night :)

Thank you for contributing!!!! 👏

**Review by uilianries — APPROVED** [review]
LGTM

File: recipes/libffi/all/conandata.yml
Comment (Croydon) [inline recipes/libffi/all/conandata.yml:13]: Please start adding `patch_description` and `patch_type`. Additionally `patch_source` if applicable

Comment (Ahajha) [reply recipes/libffi/all/conandata.yml:13]: These were from existing patches, but I can do some digging and figure out their purposes.

**Pull request merged**
